x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/LifecycleExecutionState.java

```diff
@@ -0,0 +1,253 @@
+     * {@link IndexMetaData.Builder#putCustom(String, Map)}.
+     * @return An immutable Map representation of this execution state.
+     */
+    public Map<String, String> asMap() {
+        Map<String, String> result = new HashMap<>();
+        result.put(PHASE, phase);
+        result.put(ACTION, action);
+        result.put(STEP, step);
+        result.put(FAILED_STEP, failedStep);
+        result.put(STEP_INFO, stepInfo);
+        result.put(INDEX_CREATION_DATE, String.valueOf(indexCreationDate));
+        result.put(PHASE_TIME, String.valueOf(phaseTime));
+        result.put(ACTION_TIME, String.valueOf(actionTime));
+        result.put(STEP_TIME, String.valueOf(stepTime));
+        result.put(PHASE_DEFINITION, String.valueOf(phaseDefinition));
+        return Collections.unmodifiableMap(result);
+    }
+
+    public String getPhase() {
+        return phase;
+    }
+
+    public String getAction() {
+        return action;
+    }
+
+    public String getStep() {
+        return step;
+    }
+
+    public String getFailedStep() {
+        return failedStep;
+    }
+
+    public String getStepInfo() {
+        return stepInfo;
+    }
+
+    public String getPhaseDefinition() {
+        return phaseDefinition;
+    }
+
+    public long getIndexCreationDate() {
+        return indexCreationDate;
+    }
+
+    public long getPhaseTime() {
+        return phaseTime;
+    }
+
+    public long getActionTime() {
+        return actionTime;
+    }
+
+    public long getStepTime() {
+        return stepTime;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        LifecycleExecutionState that = (LifecycleExecutionState) o;
+        return getIndexCreationDate() == that.getIndexCreationDate() &&
+            getPhaseTime() == that.getPhaseTime() &&
+            getActionTime() == that.getActionTime() &&
+            getStepTime() == that.getStepTime() &&
+            Objects.equals(getPhase(), that.getPhase()) &&
+            Objects.equals(getAction(), that.getAction()) &&
+            Objects.equals(getStep(), that.getStep()) &&
+            Objects.equals(getFailedStep(), that.getFailedStep()) &&
+            Objects.equals(getStepInfo(), that.getStepInfo()) &&
+            Objects.equals(getPhaseDefinition(), that.getPhaseDefinition());
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(getPhase(), getAction(), getStep(), getFailedStep(), getStepInfo(), getPhaseDefinition(),
+            getIndexCreationDate(), getPhaseTime(), getActionTime(), getStepTime());
+    }
+
+    public static class Builder {
+        private String phase = "";
+        private String action = "";
+        private String step = "";
+        private String failedStep = "";
+        private String stepInfo = "";
+        private String phaseDefinition = "";
```

Comment:
See my above comment, I think these should default to `null` and not be written if they aren't present